src/ImagePotentialBondGPU.cuh

```diff
@@ -0,0 +1,352 @@
+                                               const group_storage<group_size>* blist,
+                                               const Index2D blist_idx,
+                                               const unsigned int* bpos_list,
+                                               const unsigned int* n_bonds_list,
+                                               const unsigned int n_bond_type,
+                                               const typename evaluator::param_type* d_params,
+                                               unsigned int* d_flags)
+    {
+    // start by identifying which particle we are to handle
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    // shared array for per bond type parameters
+    extern __shared__ char s_data[];
+    typename evaluator::param_type* s_params = (typename evaluator::param_type*)(&s_data[0]);
+
+    if (enable_shared_cache)
+        {
+        // load in per bond type parameters
+        for (unsigned int cur_offset = 0; cur_offset < n_bond_type; cur_offset += blockDim.x)
+            {
+            if (cur_offset + threadIdx.x < n_bond_type)
+                {
+                s_params[cur_offset + threadIdx.x] = d_params[cur_offset + threadIdx.x];
+                }
+            }
+
+        __syncthreads();
+        }
+
+    if (idx >= N)
+        return;
+
+    // load in the length of the list for this thread (MEM TRANSFER: 4 bytes)
+    int n_bonds = n_bonds_list[idx];
+
+    // read in the position of our particle. (MEM TRANSFER: 16 bytes)
+    Scalar4 postype = __ldg(d_pos + idx);
+    Scalar3 pos = make_scalar3(postype.x, postype.y, postype.z);
+
+    Scalar q(0);
+    if (evaluator::needsCharge())
+        {
+        q = __ldg(d_charge + idx);
+        }
+    else
+        q += 0; // Silence compiler warning.
+
+    // initialize the force to 0
+    Scalar4 force = make_scalar4(0, 0, 0, 0);
+    // initialize the virial tensor to 0
+    Scalar virial[6];
+    for (unsigned int i = 0; i < 6; i++)
+        virial[i] = 0;
+
+    // loop over neighbors
+    for (int bond_idx = 0; bond_idx < n_bonds; bond_idx++)
+        {
+        int cur_bond_pos = bpos_list[blist_idx(idx, bond_idx)];
+
+        if (cur_bond_pos > 1)
+            continue;
+
+        group_storage<group_size> cur_bond = blist[blist_idx(idx, bond_idx)];
+
+        int cur_bond_idx = cur_bond.idx[0];
+        int cur_bond_type = cur_bond.idx[group_size - 1];
+
+        // get the bonded particle's position (MEM TRANSFER: 16 bytes)
+        Scalar4 neigh_postypej = __ldg(d_pos + cur_bond_idx);
+        Scalar3 neigh_pos = make_scalar3(neigh_postypej.x, neigh_postypej.y, neigh_postypej.z);
+
+        // get images of particles
+        const int3 img_a = d_images[idx];
+        const int3 img_b = d_images[cur_bond_idx];
+
+        // get relative vector between particles using image shift
+        const Scalar3 dx = box.shift(pos - neigh_pos, img_a - img_b);
```

Comment:
Just noting that HOOMD seems to use opposite direction for these differences on CPU and GPU (here it is `a-b`, but it is `b-a` on the CPU). That seems to be right because then the force is added below. I'm not sure what the reason is for that.